sdk/cosmos/azure-cosmos/azure/cosmos/aio/_cosmos_client_connection_async.py

```diff
@@ -2349,6 +2351,55 @@ def __GetBodiesFromQueryResult(result):
         # Query operations will use ReadEndpoint even though it uses POST(for regular query operations)
         request_params = _request_object.RequestObject(typ, documents._OperationType.SqlQuery)
         req_headers = base.GetHeaders(self, initial_headers, "post", path, id_, typ, options, partition_key_range_id)
+
+        # check if query has prefix partition key
+        cont_prop = kwargs.pop("containerProperties", None)
+        partition_key = options.get("partitionKey", None)
+        isPrefixPartitionQuery = False
+        partition_key_definition = None
+        if cont_prop:
+            cont_prop = await cont_prop()
+            pk_properties = cont_prop["partitionKey"]
+            partition_key_definition = PartitionKey(path=pk_properties["paths"], kind=pk_properties["kind"])
+            if partition_key_definition.kind == "MultiHash" and\
+                    (type(partition_key) == list and len(partition_key_definition['paths']) != len(partition_key)):
+                isPrefixPartitionQuery = True
+
+        if isPrefixPartitionQuery:
+            # here get the overlapping ranges
+            req_headers.pop(http_constants.HttpHeaders.PartitionKey, None)
+            feedrangeEPK = partition_key_definition._get_epk_range_for_prefix_partition_key(partition_key)  # cspell:disable-line # pylint: disable=line-too-long
+            over_lapping_ranges = await self._routing_map_provider.get_overlapping_ranges(id_, [feedrangeEPK])
+            results = None
+            for over_lapping_range in over_lapping_ranges:
+                # It is possible for the over lapping range to include multiple physical partitions
+                # we should return query results for all the partitions that are overlapped.
+                single_range = routing_range.Range.PartitionKeyRangeToRange(over_lapping_range)
+                if single_range.min == feedrangeEPK.min and single_range.max == feedrangeEPK.max:
+                    # The EpkRange spans exactly one physical partition
+                    # In this case we can route to the physical pk range id
+                    req_headers[http_constants.HttpHeaders.PartitionKeyRangeID] = over_lapping_range["id"]
+                else:
+                    # The EpkRange spans less than single physical partition
+                    # In this case we route to the physical partition and
+                    # pass the epk range headers to filter within partition
+                    req_headers[http_constants.HttpHeaders.PartitionKeyRangeID] = over_lapping_range["id"]
+                    req_headers[http_constants.HttpHeaders.StartEpkString] = feedrangeEPK.min
```

Comment:
This is not the feedRangeEPK.min, but rather, the min of the overlappingRange? Imagine I send a PK that spans AA-CC
And I have partition 0 with 00-BB and partition 1 with BB-DD
Then the right output would be going to partition 0 with AA-BB and then going to partition 1 with BB-CC

Reply:
But would it not still work if the pk range of AA-CC is provided to all the partitions?
for partition 0 it will start at AA and until the max of partition 0 (BB) since it is less than that of the pk (CC) and same with partition 1 where it will start at BB since its the min of partition 1 and it is greater than the min of Pk range and will go until CC since it is still within the pk range.

Reply:
The service contract as far as I know is that the EPK Range headers are always within the partition the request is for.

In this case the logic is not really complicated. If overlapping ranges returns X ranges, range 0 is feedRange.min/range.max, range X-1 (last) is range.min/feedRange.max. Anything in-between would fall into the `if` of full partition.

Reply:
Got it, that does make sense.